resolver.go

```diff
@@ -0,0 +1,391 @@
+				slice = reflect.Append(slice, reflect.ValueOf(itemValue))
+			} else {
+				// Keep going to check other array items for vailidity.
+				continue
+			}
+		}
+		pv = slice.Interface()
+	default:
+		if typ == timeType {
+			dt, err := time.Parse(timeFormat, pstr)
+			if err != nil {
+				ctx.AddError(&ErrorDetail{
+					Message:  "cannot parse time",
+					Location: location + "." + name,
+					Value:    pstr,
+				})
+				return nil
+			}
+			pv = dt
+		} else {
+			pv = pstr
+		}
+	}
+
+	return pv
+}
+
+func setFields(ctx *hcontext, req *http.Request, input reflect.Value, t reflect.Type) {
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	if input.Kind() == reflect.Ptr {
+		input = input.Elem()
+	}
+
+	if t.Kind() != reflect.Struct {
+		panic("not a struct")
+	}
+
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		inField := input.Field(i)
+
+		if f.Anonymous {
+			// Embedded struct
+			setFields(ctx, req, inField, f.Type)
+			continue
+		}
+
+		if _, ok := f.Tag.Lookup("body"); ok || f.Name == "Body" {
+			// Special case: body field is a reader for streaming
+			if f.Type == readerType {
+				inField.Set(reflect.ValueOf(req.Body))
+				continue
+			}
+
+			// Check if a content-length has been sent. If it's too big then there
+			// is no need to waste time reading.
+			if length := req.Header.Get("Content-Length"); length != "" {
+				if l, err := strconv.ParseInt(length, 10, 64); err == nil {
+					if l > ctx.op.maxBodyBytes {
+						ctx.AddError(&ErrorDetail{
+							Message:  fmt.Sprintf("Request body too large, limit = %d bytes", ctx.op.maxBodyBytes),
+							Location: "body",
+							Value:    length,
+						})
+						continue
+					}
+				}
+			}
+
+			// Load the body (read/unmarshal).
+			data, err := ioutil.ReadAll(req.Body)
+			if err != nil {
+				if strings.Contains(err.Error(), "request body too large") {
+					ctx.AddError(&ErrorDetail{
+						Message:  fmt.Sprintf("Request body too large, limit = %d bytes", ctx.op.maxBodyBytes),
+						Location: "body",
+					})
+				} else if e, ok := err.(net.Error); ok && e.Timeout() {
+					ctx.AddError(&ErrorDetail{
+						Message:  fmt.Sprintf("Request body took too long to read: timed out after %v", ctx.op.bodyReadTimeout),
+						Location: "body",
+					})
+				} else {
+					panic(err)
+				}
+				continue
+			}
+
+			if ctx.op.requestSchema != nil && ctx.op.requestSchema.HasValidation() {
+				if !validAgainstSchema(ctx, "body.", ctx.op.requestSchema, data) {
+					continue
+				}
+			}
+
+			err = json.Unmarshal(data, inField.Addr().Interface())
+			if err != nil {
+				panic(err)
+			}
+			continue
+		}
+
+		var pv string
+		var pname string
+		var location string
+		timeFormat := time.RFC3339Nano
+		if v, ok := f.Tag.Lookup("default"); ok {
+			pv = v
+		}
+
+		if name, ok := f.Tag.Lookup("path"); ok {
+			pname = name
+			location = "path"
+			if v := chi.URLParam(req, name); v != "" {
+				pv = v
+			}
+		}
```

Comment:
You have this logic repeated three times; is this a switch in disguise?

Reply:
You have to call `Lookup(...)` with a different tag name each time, so I wasn't able to make this a `switch`. Maybe I'm missing something though?

Reply:
I was apparently quite scattered when I wrote `switch` because looking at it again, I'm not sure what I meant.

I _think_ what I was trying to ask was whether we wanted a helper function to perform this function on a repeated basis, but since one of the three blocks appears likely to change in the future, on review, I'm not sure there's nothing here that needs to be done.